# Job Market Data Pipeline

**An automated Python data pipeline that performs cleansing and analysis on LinkedIn job postings data from 2023-2024. Built primarily using pandas and numpy, the pipeline uncovers hring trends, in-demand skills and salary patterns across the data industry.**

**Imports**

In [1]:
import numpy as np
import pandas as pd

Imported the two libraries pandas and numpy; pandas, being the tool used for loading, cleaning and manipulating data while numpy is used for numerical and statistical analysis. 

**Loading Data**

In [2]:
df = pd.read_csv("raw/postings.csv")

df.head(10)

,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,Requirements: \n\nWe are seeking a College or ...,1.713398e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,1.712858e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,We are currently accepting resumes for FOH - A...,1.713278e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,This position requires a baseline understandin...,1.712896e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,1.713452e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0
5,91700727,Downtown Raleigh Alliance,Economic Development and Planning Intern,Job summary:The Economic Development & Plannin...,20.0,HOURLY,"Raleigh, NC",1481176.0,9.0,NaN,...,NaN,1.713456e+12,NaN,0,INTERNSHIP,USD,BASE_SALARY,35360.0,27601.0,37183.0
6,103254301,Raw Cereal,Producer,Company DescriptionRaw Cereal is a creative de...,300000.0,YEARLY,United States,81942316.0,7.0,NaN,...,NaN,1.712861e+12,NaN,0,CONTRACT,USD,BASE_SALARY,180000.0,NaN,NaN
7,112576855,NaN,Building Engineer,Summary: Due to the pending retirement of our ...,120000.0,YEARLY,"San Francisco, CA",NaN,2.0,NaN,...,NaN,1.712443e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,105000.0,94101.0,6075.0
8,1218575,Children's Nebraska,Respiratory Therapist,"At Children’s, the region’s only full-service ...",NaN,NaN,"Omaha, NE",721189.0,3.0,NaN,...,• Requires the ability to communicate effectiv...,1.712348e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,68102.0,31055.0
9,2264355,Bay West Church,Worship Leader,It is an exciting time to be a part of our chu...,NaN,MONTHLY,"Palm Bay, FL",28631247.0,5.0,350.0,...,"Knowledge, Skills and Abilities: 1. Proficient...",1.712456e+12,NaN,0,PART_TIME,USD,BASE_SALARY,4200.0,32905.0,12009.0


In [3]:
print(df.shape)

(123849, 31)


In [4]:
print(df.columns)

Index(['job_id', 'company_name', 'title', 'description', 'max_salary',
       'pay_period', 'location', 'company_id', 'views', 'med_salary',
       'min_salary', 'formatted_work_type', 'applies', 'original_listed_time',
       'remote_allowed', 'job_posting_url', 'application_url',
       'application_type', 'expiry', 'closed_time',
       'formatted_experience_level', 'skills_desc', 'listed_time',
       'posting_domain', 'sponsored', 'work_type', 'currency',
       'compensation_type', 'normalized_salary', 'zip_code', 'fips'],
      dtype='object')


**Locating missing values**

In [5]:
print(df.isnull().sum())

job_id                             0
company_name                    1719
title                              0
description                        7
max_salary                     94056
pay_period                     87776
location                           0
company_id                      1717
views                           1689
med_salary                    117569
min_salary                     94056
formatted_work_type                0
applies                       100529
original_listed_time               0
remote_allowed                108603
job_posting_url                    0
application_url                36665
application_type                   0
expiry                             0
closed_time                   122776
formatted_experience_level     29409
skills_desc                   121410
listed_time                        0
posting_domain                 39968
sponsored                          0
work_type                          0
currency                       87776
c

**My next step was to find missing values within my dataframe using the isnull() 
function. This displays each column name and how many missing values are in those 
columns. This gives me a greater understanding of which data points aren't worth 
analysing.**

**Dropping columns**

In [6]:
df = df [['title', 'company_name', 'location', 'formatted_work_type', 'remote_allowed', 'formatted_experience_level', 'min_salary', 'max_salary', 'normalized_salary', 'views', 'applies', 'listed_time']]

print(df.shape)

(123849, 12)


**Here, I have selected only the relevant columns for analysis and dropped the rest 
from the dataframe. I identified which columns to remove by using the isnull().sum() 
function to check for missing values. Columns with excessive missing values such as 
closed_time and skills_desc were removed, as were irrelevant columns such 
as job_id, posting_domain and application_url.**

**Checking data types**

In [7]:
df.dtypes

title                          object
company_name                   object
location                       object
formatted_work_type            object
remote_allowed                float64
formatted_experience_level     object
min_salary                    float64
max_salary                    float64
normalized_salary             float64
views                         float64
applies                       float64
listed_time                   float64
dtype: object

**Here, I have checked the data types of each column to ensure they are accurate 
and best suited for analysis. Title, company_name and location are all object 
(string/text) columns which is correct. However, columns such as min_salary, 
views and applies are float64 (decimal) which is not ideal — these will be 
converted to integers as you cannot have half a salary or half a view. 
Additionally, remote_allowed will be converted to a boolean (True/False) and 
listed_time will be converted to a datetime format.**

**Converting data types**

In [8]:
df = df.fillna(0).astype({'min_salary':'int','max_salary':'int','normalized_salary':'int','views':'int','applies':'int'})

**Here, my aim was to convert several column data types from float64 to integers. 
These columns were min_salary, max_salary, normalized_salary, views and applies. 
Before converting, I first had to handle the NULL values within these columns by 
filling them with the value '0' using the fillna() method, chained with astype() to perform 
the conversion in a single line. Skipping this step would result in an error, as 
Python cannot convert NULL values to another data type.**

**Converting column to datetime**

In [9]:
df['listed_time'] = pd.to_datetime(df['listed_time'], unit = 'ms')

In [10]:
df['listed_time']

0        2024-04-17 23:45:08
1        2024-04-11 17:51:27
2        2024-04-16 14:26:54
3        2024-04-12 04:23:32
4        2024-04-18 14:52:23
                 ...        
123844   2024-04-20 00:00:23
123845   2024-04-20 00:17:16
123846   2024-04-20 00:18:59
123847   2024-04-20 00:23:52
123848   2024-04-20 00:23:36
Name: listed_time, Length: 123849, dtype: datetime64[ns]

**Here, I converted the listed_time column from a Unix timestamp to a readable 
datetime format. During this step I encountered a debugging issue — the raw values 
were stored as scientific numbers (e.g. 1.71e+12), representing milliseconds since 
1st January 1970. Initially the dates were converting incorrectly, displaying 1970 
dates. By inspecting the raw values I identified the issue and resolved it by using 
pd.to_datetime() with the unit='ms' argument, which correctly converted the values 
to 2023-2024 dates.**

**Checking for and dropping duplicate rows**

In [11]:
print(sum(df.duplicated()))

803


In [12]:
df = df.drop_duplicates()

In [13]:
df.shape

(123046, 12)

**Final check before analysis**

In [14]:
df

,title,company_name,location,formatted_work_type,remote_allowed,formatted_experience_level,min_salary,max_salary,normalized_salary,views,applies,listed_time
0,Marketing Coordinator,Corcoran Sawyer Smith,"Princeton, NJ",Full-time,0.0,0,17,20,38480,20,2,2024-04-17 23:45:08
1,Mental Health Therapist/Counselor,0,"Fort Collins, CO",Full-time,0.0,0,30,50,83200,1,0,2024-04-11 17:51:27
2,Assitant Restaurant Manager,The National Exemplar,"Cincinnati, OH",Full-time,0.0,0,45000,65000,55000,8,0,2024-04-16 14:26:54
3,Senior Elder Law / Trusts and Estates Associat...,"Abrams Fensterman, LLP","New Hyde Park, NY",Full-time,0.0,0,140000,175000,157500,16,0,2024-04-12 04:23:32
4,Service Technician,0,"Burlington, IA",Full-time,0.0,0,60000,80000,70000,3,0,2024-04-18 14:52:23
...,...,...,...,...,...,...,...,...,...,...,...,...
123844,Title IX/Investigations Attorney,Lozano Smith,"Walnut Creek, CA",Full-time,0.0,Mid-Senior level,120000,195000,157500,1,0,2024-04-20 00:00:23
123845,"Staff Software Engineer, ML Serving Platform",Pinterest,United States,Full-time,1.0,Mid-Senior level,0,0,0,3,0,2024-04-20 00:17:16
123846,"Account Executive, Oregon/Washington",EPS Learning,"Spokane, WA",Full-time,1.0,Mid-Senior level,0,0,0,3,0,2024-04-20 00:18:59
123847,Business Development Manager,Trelleborg Applied Technologies,"Texas, United States",Full-time,1.0,0,0,0,0,4,0,2024-04-20 00:23:52


**Changing the integers ("0") to "Unknown" in our string/text columns**

In [15]:
df['company_name'] = df['company_name'].replace(0,'Unknown')
df['location'] = df['location'].replace(0,'Unknown')
df['formatted_experience_level'] = df['formatted_experience_level'].replace(0,'Unknown')

**Converting remote_allowed column to boolean**

In [16]:
df['remote_allowed'] = df['remote_allowed'].astype('bool')

In [17]:
df["remote_allowed"].dtype

dtype('bool')

In [18]:
df.head(10)

,title,company_name,location,formatted_work_type,remote_allowed,formatted_experience_level,min_salary,max_salary,normalized_salary,views,applies,listed_time
0,Marketing Coordinator,Corcoran Sawyer Smith,"Princeton, NJ",Full-time,False,Unknown,17,20,38480,20,2,2024-04-17 23:45:08
1,Mental Health Therapist/Counselor,Unknown,"Fort Collins, CO",Full-time,False,Unknown,30,50,83200,1,0,2024-04-11 17:51:27
2,Assitant Restaurant Manager,The National Exemplar,"Cincinnati, OH",Full-time,False,Unknown,45000,65000,55000,8,0,2024-04-16 14:26:54
3,Senior Elder Law / Trusts and Estates Associat...,"Abrams Fensterman, LLP","New Hyde Park, NY",Full-time,False,Unknown,140000,175000,157500,16,0,2024-04-12 04:23:32
4,Service Technician,Unknown,"Burlington, IA",Full-time,False,Unknown,60000,80000,70000,3,0,2024-04-18 14:52:23
5,Economic Development and Planning Intern,Downtown Raleigh Alliance,"Raleigh, NC",Internship,False,Unknown,14,20,35360,9,4,2024-04-18 16:01:39
6,Producer,Raw Cereal,United States,Contract,True,Unknown,60000,300000,180000,7,1,2024-04-11 18:43:39
7,Building Engineer,Unknown,"San Francisco, CA",Full-time,False,Unknown,90000,120000,105000,2,0,2024-04-06 22:44:12
8,Respiratory Therapist,Children's Nebraska,"Omaha, NE",Full-time,False,Unknown,0,0,0,3,0,2024-04-05 20:21:40
9,Worship Leader,Bay West Church,"Palm Bay, FL",Part-time,False,Unknown,0,0,4200,5,0,2024-04-07 02:12:35


**As you can see, I've converted the remote_allowed column to boolean instead of float64. Previously showing either 0.0 or 1.0, the column values now show either True or False.**

In [19]:
print(df.shape)
print(df.dtypes)
df.head()

(123046, 12)
title                                 object
company_name                          object
location                              object
formatted_work_type                   object
remote_allowed                          bool
formatted_experience_level            object
min_salary                             int64
max_salary                             int64
normalized_salary                      int64
views                                  int64
applies                                int64
listed_time                   datetime64[ns]
dtype: object


,title,company_name,location,formatted_work_type,remote_allowed,formatted_experience_level,min_salary,max_salary,normalized_salary,views,applies,listed_time
0,Marketing Coordinator,Corcoran Sawyer Smith,"Princeton, NJ",Full-time,False,Unknown,17,20,38480,20,2,2024-04-17 23:45:08
1,Mental Health Therapist/Counselor,Unknown,"Fort Collins, CO",Full-time,False,Unknown,30,50,83200,1,0,2024-04-11 17:51:27
2,Assitant Restaurant Manager,The National Exemplar,"Cincinnati, OH",Full-time,False,Unknown,45000,65000,55000,8,0,2024-04-16 14:26:54
3,Senior Elder Law / Trusts and Estates Associat...,"Abrams Fensterman, LLP","New Hyde Park, NY",Full-time,False,Unknown,140000,175000,157500,16,0,2024-04-12 04:23:32
4,Service Technician,Unknown,"Burlington, IA",Full-time,False,Unknown,60000,80000,70000,3,0,2024-04-18 14:52:23


**Within Stage 1, I completed the initial cleaning process of the raw LinkedIn 
job postings dataset. This involved loading the dataset and inspecting the column 
names, data types and missing values. Irrelevant and incomplete columns were dropped 
from 31 down to 12. Null values were handled using fillna() and incorrect 0 values 
in text columns were replaced with "Unknown". Data types were converted to their 
appropriate formats including integers, boolean and datetime. Finally, 803 duplicate 
rows were removed, leaving a cleaned dataset of 123,046 rows.**

In [21]:
df.to_csv("processed/cleaned_postings.csv", index=False)